azure-kusto-data/azure/kusto/data/_cloud_settings.py

```diff
@@ -1,35 +1,79 @@
 import os
+from functools import lru_cache
+from typing import Optional
+from urllib.parse import urljoin
+
+import requests
+
+from azure.kusto.data.exceptions import KustoServiceError
 
 AUTH_ENV_VAR_NAME = "AadAuthorityUri"
 KUSTO_CLIENT_APP_ID = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
 PUBLIC_LOGIN_URL = "https://login.microsoftonline.com"
+REDIRECT_URI = "https://microsoft/kustoclient"
+KUSTO_SERVICE_RESOURCE_ID = "https://kusto.kusto.windows.net"
+FIRST_PARTY_AUTHORITY_URL = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a"
 
 
 class CloudInfo:
     """This class holds the data for a specific cloud instance."""
 
-    def __init__(self, auth_endpoint: str, kusto_client_app_id: str, redirect_uri: str):
-        self.aad_authority_uri = auth_endpoint
+    def __init__(
+        self,
+        login_endpoint: str,
+        login_mfa_required: bool,
+        kusto_client_app_id: str,
+        kusto_client_redirect_uri: str,
+        kusto_service_resource_id: str,
+        first_party_authority_url: str,
+    ):
+        self.login_endpoint = login_endpoint
+        self.login_mfa_required = login_mfa_required
         self.kusto_client_app_id = kusto_client_app_id
-        self.login_redirect_uri = redirect_uri  # will be used for interactive login
+        self.kusto_client_redirect_uri = kusto_client_redirect_uri  # will be used for interactive login
+        self.kusto_service_resource_id = kusto_service_resource_id
+        self.first_party_authority_url = first_party_authority_url
+
+    def authority_uri(self, authority_id: Optional[str]):
+        return self.login_endpoint + "/" + (authority_id or "organizations")
 
 
 class CloudSettings:
     """This class holds data for all cloud instances, and returns the specific data instance by parsing the dns suffix from a URL"""
 
-    _cloud_info = None
+    METADATA_ENDPOINT = "v1/rest/auth/metadata"
 
-    @classmethod
-    def _init_once(cls):
-        if cls._cloud_info is not None:
-            return
+    _cloud_info = None
 
-        # todo: replace this with a call to the auth metadata endpoint
-        aad_authority_uri = os.environ.get(AUTH_ENV_VAR_NAME, PUBLIC_LOGIN_URL)
-        cls._cloud_info = CloudInfo(aad_authority_uri, KUSTO_CLIENT_APP_ID, None)
+    DEFAULT_CLOUD = CloudInfo(
+        login_endpoint=os.environ.get(AUTH_ENV_VAR_NAME, PUBLIC_LOGIN_URL),
+        login_mfa_required=False,
+        kusto_client_app_id=KUSTO_CLIENT_APP_ID,
+        kusto_client_redirect_uri=REDIRECT_URI,
+        kusto_service_resource_id=KUSTO_SERVICE_RESOURCE_ID,
+        first_party_authority_url=FIRST_PARTY_AUTHORITY_URL,
+    )
 
     @classmethod
-    def get_cloud_info(cls) -> CloudInfo:
-        """Get the details of a cloud according to the DNS suffix of the provided connection string"""
-        cls._init_once()
-        return cls._cloud_info
+    @lru_cache(maxsize=None)

```

Comment:
How does this sorcery work

Reply:
It's a cool decorator from functools, it basically wraps the function and saves the result to a dict.

I've now decided to remove it, because it's impossible to control manually, which is needed for tests.
So now it's a simple dict+lock